better-code/src/chapter-2-contracts.md

```diff
@@ -0,0 +1,1129 @@
+> whose interaction is based on precisely defined specifications of
+> the mutual obligations — contracts.*
+>
+>   —Building bug-free O-O software: An Introduction to Design by Contract™
+>    https://www.eiffel.com/values/design-by-contract/introduction/
+
+In the mid 1980s, The French computer scientist Bertrand Meyer took
+Hoare Logic, and shaped it into a practical discipline for software
+engineering that he called “Design By Contract.”  Meyer was focused on
+software components—types and methods.  He realized that there's
+special power in the Hoare triple that captures the intentions of a
+function's author—the function's **contract**:
+
+- The precondition describes which calls to a function should be
+  considered valid.
+- The postcondition omits anything that is purely an artifact of the
+  particular implementation, such as what happens when an invalid call
+  is made.
+- It is general—describing the result for all inputs the author
+  intends to support—so it can be applied in reasoning about any call
+  to the function.
+- An author can evolve the implementation so that valid uses by
+  clients stay valid and retain their meaning.
+
+Contracts describe the rules that govern how one piece of software
+talks to another. In other words, they're relationships.  Thinking in
+terms of relationships is one of the themes of Better Code, and you
+can expect us to point relationships out as they come up in our
+material.
+
+### Which code is to blame?
+
+When something goes wrong in software, focusing on which *person* to
+blame is counterproductive, but deciding which *code* is to blame is
+the first step.  Contracts tell us which code needs fixing:
+
+- If preconditions aren't satisifed, that's a bug in the caller.  The
+  function is not required to make any promises[^no-promises] in that case.
+
+- If preconditions are statisfied but postconditions are not
+  fulfilled, that's a bug in the callee, or in something it calls.
+
+[^no-promises]: In fact, a function *shouldn't* make any promises in
+  case of precondition violation, because then the caller has a right
+  to call the function and expect the promised result: it's no longer
+  a precondition violation.  There is one exception, especially in a
+  safe-by-default language like Swift: all functions not explicitly
+  labeled unsafe implicitly promise not to have arbitrary “undefined
+  behavior” when their preconditions are violated.  But that's a
+  blanket guarantee that you don't spell out in contracts.
+
+And if you ever find yourself with something that's clearly a bug but
+you can't decide which code to blame, that's a sign that a contract is
+missing or incomplete.  A system that exposes these situations, where
+everybody is playing by the rules but things still go wrong, is known
+by the technical term “footgun.”
+
+### Type Invariants
+
+The other contribution of Meyer's Design by Contract was to apply the
+idea of “invariants” to user-defined types with private parts.  A
```

Comment:
Does user-defined add anything here? Implementation defined types can have invariants.

Reply:
```suggestion
idea of “invariants” to types with private parts.  A
```

Reply:
Another place where "implementation" may have added confusion. Clearly std:: types have private parts, and they're not user-defined. Or, is the standard library a user (of the language) in this case? I don't think "private parts" is superfluous. But I do think "user-defined" is.

Reply:
But a type without visible private parts can still have an invariant.  I think I have to nix both of those qualifiers.